### PR TITLE
Microbatch parallelism

### LINKED ISSUE #10853: [Feature] automatically detect which batches can not be run in parallel
### Is this your first time submitting a feature request?

- [X] I have read the [expectations for open source contributors](https://docs.getdbt.com/docs/contributing/oss-expectations)
- [X] I have searched the existing issues, and I could not find an existing issue for this feature
- [X] I am requesting a straightforward extension of existing dbt functionality, rather than a Big Idea better suited to a discussion

### Describe the feature

can we auto-detect which microbatch models cannot run their batches in parallel? 
- those that use `{{ this }}`
- etc.

### Describe alternatives you've considered

_No response_

### Who will this benefit?

_No response_

### Are you interested in contributing this feature?

_No response_

### Anything else?

_No response_

### LINKED ISSUE #10855: [Feature] microbatches should default run in parallel according to threads
### Describe the feature

For faster processing of microbatch models, we should by default run the batches in parallel. We are assuming batches can be run independently _unless_ the adapter doesn't support it.

We could additionally:
- raise an exception when the adapter can't run batches in parallel but the microbatch model is configured as such
- have a config to optionally run batches _sequentially_ for specific use cases
- [automatically detect which batches can not be run in parallel](https://github.com/dbt-labs/dbt-core/issues/10853)

[Slack thread](https://dbt-labs.slack.com/archives/C078A5CBAKH/p1726522615227489)

### PR DESCRIPTION
**Pull request by MichelleArk**
Resolves #10853
Resolves #10855

### Problem

Part of the benefit of microbatch, is updating a model in smaller (micro) batches. However, before the work in this PR, those batches were run _sequentially_. Having them run sequentially didn't negate all benefits of splitting into batches (partial instead of complete failures, better data guarantees, easier backfills, etc), however because they were running serially there weren't any speed gains. We wanted the ability to run batches in parallel. This PR does that.

### Solution

First we pulled all the batch execution logic into a new `MicrobatchModelRunner`, which is called from the `ModelRunner` when a microbatch model is encountered. Handwaving all the complexities of creating a new <X>Runner class, the class determines whether things can be run in parallel in the following.

First, if any of the following are true, a batch is run _sequentially_
1. The adapter doesn't support concurrent batches
2. The relation in the data warehouse for the model doesn't exist

If neither (1) nor (2) was hit, then we check if a config `concurrent_batches` is set for the model. If the value for that config is `True` then we run the batches in parallel, if `False` we run the batches sequentially.

If however `concurrent_batches` is `None` (i.e. not set), then we check if the model jinja contains a reference to `this`. If it references `this` then we run the batches sequentially. Otherwise, we run them in parallel. 

### Checklist

- [X] I have read [the contributing guide](https://github.com/dbt-labs/dbt-core/blob/main/CONTRIBUTING.md) and understand what's expected of me.
- [x] I have run this code in development, and it appears to resolve the stated issue.
- [X] This PR includes tests, or tests are not required or relevant for this PR.
- [X] This PR has no interface changes (e.g., macros, CLI, logs, JSON artifacts, config files, adapter interface, etc.) or this PR has already received feedback and approval from Product or DX.
- [X] This PR includes [type annotations](https://docs.python.org/3/library/typing.html) for new and modified functions.

### CODE REVIEW

**Comment by QMalcolm** [comment]
Related: https://github.com/dbt-labs/schemas.getdbt.com/pull/71

File: dev-requirements.txt
Comment (QMalcolm) [inline dev-requirements.txt:1]: Need to rever prior to merge

**Comment by QMalcolm** [comment]
All tests are passing. We need 3 things to happen before we can merge.
1. Merge https://github.com/dbt-labs/dbt-adapters/pull/358
2. Revert [8bf1504](https://github.com/dbt-labs/dbt-core/pull/10958/commits/8bf1504d478ee45550df5032a3d366545b9f2947)
3. Merge https://github.com/dbt-labs/schemas.getdbt.com/pull/71

**Review by gshank — APPROVED** [review]

File: core/dbt/task/run.py
Comment (MichelleArk) [inline core/dbt/task/run.py:374]: let's remove the `on threading.current_thread().name` here, it was just for development

File: core/dbt/task/run.py
Comment (MichelleArk) [inline core/dbt/task/run.py:263]: This can probably move to the MicrobatchModelRunner entirely

File: core/dbt/task/run.py
Comment (MichelleArk) [reply core/dbt/task/run.py:374]: Hmm, this is probably going to clutter up the logs for users. I wonder if we get any information about utilized threads in the debug logs we can leverage instead for the purposes of testing

Comment (MichelleArk) [reply core/dbt/task/run.py:374]: What do you think about extending the debug `NodeStart` with a `thread_name` and passing the thread_name through GraphRunnableTask.call_runner (where NodeStart is currently fired)

Comment (QMalcolm) [reply core/dbt/task/run.py:374]: 🤔 Let's try it! We're still waiting on the approval of the dbt-adapters PR anyway

**Comment by QMalcolm** [comment]
@cla-bot check

**Comment by QMalcolm** [comment]
@cla-bot check

**Comment by QMalcolm** [comment]
@cla-bot check

**Comment by QMalcolm** [comment]
The cla-bot is refusing to show up 🤔

**Comment by QMalcolm** [comment]
@cla-bot check

**Comment by QMalcolm** [comment]
@cla-bot check

File: core/dbt/events/core_types_pb2.py
Comment (QMalcolm) [reply core/dbt/events/core_types_pb2.py:4]: 🤔

**Comment by MichelleArk** [comment]
Merging and bypassing branch protection for `verification/cla-signed`, as the check seems broken across other branches as well.

**Pull request merged**
